shell/browser/native_window.h

```diff
@@ -425,6 +427,9 @@ class NativeWindow : public base::SupportsUserData,
   // throttling, then throttling in the `ui::Compositor` will be disabled.
   void UpdateBackgroundThrottlingState();
 
+  void SaveWindowState();
+  void DoSaveWindowState();
```

Comment:
```suggestion
  void ScheduleSaveWindowState();
  void SaveWindowState();
  void FlushWindowState();
```

Schedule should do what you currently do, set up the timer.  Save should be your current code. Then `FlushWindowState` is a new method that calls `save_window_state_timer_->FireNow()` which will avoid duplicate calls and ensure that the timer is flushed rather than just calling the method in addition to having a pending timer.

Reply:
Also duplicate calls won't occur with the current implementation. The OneShotTimer::Start() method automatically cancels any pending timer before starting a new one.

Reply:
There is a race present here because of this direct call to `SaveWindowState` --> https://github.com/electron/electron/pull/47425/files#diff-9e3935b7f7b31cb73a8e976e9725bab5c20c499dc5d1e33c7c7f62e3f90b72c9R181 during window close, if a timer was previously scheduled it will still be fired.

What we actually want on the line I linked is to flush any pending timer, not run the save handler directly (because that might be work we don't need to do, and in it's current impl can result in double calls)

Reply:
Added FlushWindowState in latest commit